% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/process_ser.R
\name{process_ser}
\alias{process_ser}
\title{Processes counts into a Seurat object prepared for alignment.
Parameters:}
\usage{
process_ser(
  ser,
  mt_handle = NULL,
  scale_umi = TRUE,
  g2m_genes = NULL,
  s_genes = NULL,
  res = 0.8,
  other_sets = NULL,
  ref_ser = NULL
)
}
\arguments{
\item{ser}{Seurat object to process.}

\item{mt_handle}{Regex used to identify mitochondrial genes for scaling. If 
left blank mt gene % will not be used to scale.}

\item{scale_umi}{Whether or not to scale on total UMI count}

\item{g2m_genes}{Genes to use for g2m scoring and scaling. If left blank
cell cycle scoring and scaling will not be done.}

\item{s_genes}{Genes to use for s scoring and scaling. If left blank
cell cycle scoring and scaling will not be done.}

\item{res}{Resolution for clustering}

\item{other_sets}{A named list of gene sets to be used similar to %mt for 
scoring and scaling. Names will appear in metadata.}

\item{ref_ser}{A processed reference Seurat object used to as reference
                     for cell selection.


Reads in a blank ser object (usually from align_sers.R) and processes
with a traditional Seurat pipeline. By default will scale both RNA and
the default assay if not RNA but will perform PCA on default assay.}
}
\value{
Outputs a processed Seurat outputs (PCA, UMAP, Phate)
}
\description{
Processes counts into a Seurat object prepared for alignment.
Parameters:
}
